File: ios-native/KyouNoOgatore/RecordCore/Sources/RecordCore/KyonoTransfer.swift
import Foundation

// ネイティブ移植 Step 3(マスタープラン§2-3): index.html:2054 buildExportString / index.html:2082
// importData の1:1移植(UIのconfirm/alert/location.reloadを除いたデータ変換部分のみ)。
// これがPWA版のエクスポート文字列をそのままネイティブへの引っ越しインポート経路にする土台(§2-3)。
public enum KyonoTransferError: Error { case invalidFormat }

public enum KyonoTransfer {
    private static let prefix = "KYONO1:"
    private static let maxImportLength = 300_000
    private static let maxKeys = 50
    private static let maxValueLength = 200_000

    // "KYONO1:" + Base64(UTF8 bytes of JSON{v:1,data:{kyono_*: 生JSON文字列}})。
    // JS版はbtoaがLatin1専用のため unescape(encodeURIComponent(...)) でUTF8バイト列に変換してから
    // btoaしている。Swiftはbase64EncodedString()がバイト列を直接扱えるので、そのトリックは不要
    // (結果として得られるbase64は同じ「UTF8バイト列のbase64」なのでJS版と相互にデコード可能)。
    public static func buildExportString(_ store: RecordStore) -> String {
        let data = store.allRawKyonoEntries
        let envelope: [String: Any] = ["v": 1, "data": data]
        let jsonData = try! JSONSerialization.data(withJSONObject: envelope, options: [.sortedKeys])
        return prefix + jsonData.base64EncodedString()
    }

    // 検証(プレフィックス→Base64→JSON→中身)が全部通ってから書き込む。壊れた文字列では何も変更しない
    // (index.html:2081のコメントと同じ方針)。呼び出し側のconfirm確認・「よみこみました」表示・
    // location.reload相当はUI層(Step 5c以降)の責務でありここには含めない。
    public static func importString(_ str: String, into store: RecordStore) throws {
        guard str.hasPrefix(prefix), str.count <= maxImportLength else {
            throw KyonoTransferError.invalidFormat
        }
        let b64 = String(str.dropFirst(prefix.count))
        guard let decoded = Data(base64Encoded: b64),
              let obj = try? JSONSerialization.jsonObject(with: decoded)
        else { throw KyonoTransferError.invalidFormat }

        let src: [String: Any]
        if let envelope = obj as? [String: Any],
           let v = envelope["v"] as? NSNumber, v.intValue >= 1,
           let data = envelope["data"] as? [String: Any] {
            src = data
        } else if let flat = obj as? [String: Any] {
            src = flat // v無し(旧形式): トップレベルそのものがdata。index.html:2090と同じ両対応
        } else {
            throw KyonoTransferError.invalidFormat
        }

        var accepted: [String: String] = [:]
        for (k, v) in src {
            if accepted.count >= maxKeys { break } // アプリが使うキーは約20個・50は十分な上限
            guard k.hasPrefix("kyono_") else { continue }
            guard let s = v as? String, s.count <= maxValueLength else { continue }
            accepted[k] = s
        }
        guard !accepted.isEmpty else { throw KyonoTransferError.invalidFormat }

        for (k, v) in accepted { store.setRaw(k, v) }

        // 取り込んだキー＋端末の表示設定 以外の既存kyono_キーを掃除(index.html:2100-2104)。
        // 書込みを先に済ませてから掃除するので、途中で失敗しても既存記録は消えない。
        var keep = Set(accepted.keys)
        keep.insert("kyono_theme")
        keep.insert("kyono_bigtext")
        for k in store.allRawKyonoEntries.keys where !keep.contains(k) {
            store.removeRaw(k)
        }
    }
}

File: android-native/KyouNoOgatore/app/src/main/java/jp/ogatore/kyouno/record/KyonoTransfer.kt
package jp.ogatore.kyouno.record

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonObject
import java.util.Base64

// ネイティブ移植 Step 3(マスタープラン§2-3): index.html:2054 buildExportString / index.html:2082
// importData の1:1移植(UIのconfirm/alert/location.reloadを除いたデータ変換部分のみ)。
// これがPWA版のエクスポート文字列をそのままネイティブへの引っ越しインポート経路にする土台(§2-3)。
class KyonoTransferException(message: String) : Exception(message)

object KyonoTransfer {
    private const val PREFIX = "KYONO1:"
    private const val MAX_IMPORT_LENGTH = 300_000
    private const val MAX_KEYS = 50
    private const val MAX_VALUE_LENGTH = 200_000

    // "KYONO1:" + Base64(UTF8 bytes of JSON{v:1,data:{kyono_*: 生JSON文字列}})。
    // JS版はbtoaがLatin1専用のため unescape(encodeURIComponent(...)) でUTF8バイト列に変換してから
    // btoaしている。java.util.Base64はバイト列を直接扱えるので、そのトリックは不要
    // (結果として得られるbase64は同じ「UTF8バイト列のbase64」なのでJS版と相互にデコード可能)。
    fun buildExportString(store: RecordStore): String {
        val data = store.allRawKyonoEntries
        val envelope = buildJsonObject {
            put("v", 1)
            putJsonObject("data") { for ((k, v) in data) put(k, v) }
        }
        return PREFIX + Base64.getEncoder().encodeToString(envelope.toString().toByteArray(Charsets.UTF_8))
    }

    // 検証(プレフィックス→Base64→JSON→中身)が全部通ってから書き込む。壊れた文字列では何も変更しない
    // (index.html:2081のコメントと同じ方針)。呼び出し側のconfirm確認・「よみこみました」表示・
    // 再読込相当はUI層(Step 5c以降)の責務でありここには含めない。
    fun importString(str: String, store: RecordStore) {
        if (!str.startsWith(PREFIX) || str.length > MAX_IMPORT_LENGTH) {
            throw KyonoTransferException("invalid prefix or too long")
        }
        val b64 = str.substring(PREFIX.length)
        val decoded = runCatching { Base64.getDecoder().decode(b64) }.getOrNull()
            ?: throw KyonoTransferException("invalid base64")
        val text = String(decoded, Charsets.UTF_8)
        val elem = runCatching { Json.parseToJsonElement(text) }.getOrNull()
            ?: throw KyonoTransferException("invalid json")
        val obj = elem as? JsonObject ?: throw KyonoTransferException("invalid json")

        val v = (obj["v"] as? JsonPrimitive)?.intOrNull
        val dataObj = obj["data"] as? JsonObject
        val src: JsonObject = if (v != null && v >= 1 && dataObj != null) dataObj else obj // v無し(旧形式)両対応

        val accepted = mutableMapOf<String, String>()
        for ((k, v2) in src) {
            if (accepted.size >= MAX_KEYS) break // アプリが使うキーは約20個・50は十分な上限
            if (!k.startsWith("kyono_")) continue
            val prim = v2 as? JsonPrimitive
            if (prim == null || !prim.isString) continue
            if (prim.content.length > MAX_VALUE_LENGTH) continue
            accepted[k] = prim.content
        }
        if (accepted.isEmpty()) throw KyonoTransferException("no valid kyono_ keys")

        for ((k, v2) in accepted) store.setRaw(k, v2)

        // 取り込んだキー＋端末の表示設定 以外の既存kyono_キーを掃除(index.html:2100-2104)。
        // 書込みを先に済ませてから掃除するので、途中で失敗しても既存記録は消えない。
        val keep = accepted.keys.toMutableSet()
        keep.add("kyono_theme")
        keep.add("kyono_bigtext")
        for (k in store.allRawKyonoEntries.keys.toList()) if (k !in keep) store.removeRaw(k)
    }
}
